### PR TITLE
Tree sitter extractor: Proper handling of `LGTM_INDEX_FILTERS`

### PR DESCRIPTION
**Pull request by RasmusWL**
If someone had used `LGTM_INDEX_FILTERS=exclude:**/*\ninclude:*.rb`
before, we would have mistakenly excluded all files :|
(LGTM_INDEX_FILTERS is a prioritized list where later matches take
priority over earlier ones)

This change is needed to support adding `exclude:**/*` as the first
filter if `paths` include a glob, which currently causes bad behavior in
the Python extractor. However, we can first introduce that change once
this PR has been merged.

I realize this change can cause more folders and files to be traversed
(since they are not just skipped with --exclude). We plan to make a
better long term fix which should bring back the previous performance.

### CODE REVIEW

**Comment by aibaars** [comment]
Q4 for QL is failing. Might be related to the change in this PR because that check excludes some paths using LGTM_INDEX_FILTERS

**Comment by RasmusWL** [comment]
looks that way yes. I'll do some debugging :+1: https://github.com/github/codeql/blob/03771ffad2c1fe4de73258e3ca23b81266da033e/.github/workflows/ql-for-ql-build.yml#L63-L66

**Comment by RasmusWL** [comment]
I had not thought it through properly in terms of the tree-sitter extractor and `--also-match`, so had to make another commit with a fix. It outlines the 3 different cases where filters are present. Although this is a bit tricky, I hope reviewers will be easily convinced the behavior is the same as before.

**Comment by RasmusWL** [comment]
Ah yes, I should have mentioned that I struggled a bit with the rust code. I'll admit I'm not super proficient, so couldn't get rid of the `tmp` assignment. I'm more than happy to change this bit of the code if you have better solutions :blush:

File: shared/tree-sitter-extractor/src/autobuilder.rs [hunk 1]
```diff
@@ -74,14 +74,35 @@ impl Autobuilder {
...
+        let lgtm_index_filters = tmp.split('\n');
+        let lgtm_index_filters_has_include = lgtm_index_filters
+            .clone()
+            .any(|s| s.starts_with("include:"));
+        let lgtm_index_filters_has_exclude = lgtm_index_filters
+            .clone()
+            .any(|s| s.starts_with("exclude:"));
+
+        if !lgtm_index_filters_has_include && lgtm_index_filters_has_exclude {
+            cmd.arg("--also-match=**/*");
```
Comment (henrymercer) [inline shared/tree-sitter-extractor/src/autobuilder.rs:98]: I believe `--also-match=**` is equivalent

**Review by henrymercer — COMMENTED** [review]
This makes sense to me, but I'd like someone more familiar with the tree-sitter extractor to approve.

**Review by hmac — APPROVED** [review]
This LGTM, though I'm not too familiar with the behaviour of the `--also-match` flags to `codeql index-files`.

**Pull request merged**